File: python/1630.arithmetic-subarrays.py
class Solution:
    def checkArithmeticSubarrays(self, nums: List[int], l: List[int], r: List[int]) -> List[bool]:

        def is_arithmetic(arr: List[int]) -> bool:
            lo, hi = min(arr), max(arr)

            if (hi - lo) % (len(arr) - 1) != 0:
                return False
            
            diff = (hi - lo) // (len(arr) - 1)

            if diff == 0:
                if arr != [arr[0]] * len(arr):
                    return False
                return True

            check = [False] * len(arr)
            for n in arr:
                if (n - lo) % diff != 0:
                    return False
                check[(n - lo) // diff] = True
            
            return all(check)

        res = [False] * len(min(l, r))
        for i, (left, right) in enumerate(zip(l, r)):
            res[i] = is_arithmetic(nums[left: right + 1])

        return res

File: cpp/1630.arithmetic-subarrays.cpp
class Solution {
private:
    bool check(vector<int> arr) {
        sort(arr.begin(), arr.end());
        for (int i = 1; i < arr.size() - 1; i++) {
            if (arr[i - 1] - arr[i] != arr[i] - arr[i + 1]) {
                return false;
            }
        }
        return true;
    }
public:
    vector<bool> checkArithmeticSubarrays(vector<int>& nums, vector<int>& l, vector<int>& r) {
        int n = min(l.size(), r.size());

        vector<bool> res(n, false);
        for (int i = 0; i < n; i++) {
            res[i] = check(vector<int>(nums.begin() + l[i], nums.begin() + r[i] + 1));
        }

        return res;
    }
};